creator-node/src/services/stateMachineManager/CNodeHealthManager.ts

```diff
@@ -0,0 +1,318 @@
+    )
+  }
+
+  // Check for sufficient memory space
+  const { usedMemory, totalMemory } = verboseHealthCheckResp
+  if (
+    usedMemory &&
+    totalMemory &&
+    totalMemory - usedMemory <= MINIMUM_MEMORY_AVAILABLE
+  ) {
+    throw new Error(
+      `Running low on memory=${
+        totalMemory - usedMemory
+      }bytes remaining. Minimum memory required=${MINIMUM_MEMORY_AVAILABLE}bytes`
+    )
+  }
+
+  // Check for sufficient file descriptors space
+  const { allocatedFileDescriptors, maxFileDescriptors } =
+    verboseHealthCheckResp
+  if (
+    allocatedFileDescriptors &&
+    maxFileDescriptors &&
+    allocatedFileDescriptors / maxFileDescriptors >=
+      MAX_FILE_DESCRIPTORS_ALLOCATED_PERCENTAGE
+  ) {
+    throw new Error(
+      `Running low on file descriptors availability=${
+        (allocatedFileDescriptors / maxFileDescriptors) * 100
+      }% used. Max file descriptors allocated percentage allowed=${
+        MAX_FILE_DESCRIPTORS_ALLOCATED_PERCENTAGE * 100
+      }%`
+    )
+  }
+
+  // Check historical sync data for current day
+  const { dailySyncSuccessCount, dailySyncFailCount } = verboseHealthCheckResp
+  if (
+    dailySyncSuccessCount &&
+    dailySyncFailCount &&
+    dailySyncSuccessCount + dailySyncFailCount > MINIMUM_DAILY_SYNC_COUNT &&
+    dailySyncSuccessCount / (dailySyncFailCount + dailySyncSuccessCount) <
+      MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE
+  ) {
+    throw new Error(
+      `Latest daily sync data shows that this node fails at a high rate of syncs. Successful syncs=${dailySyncSuccessCount} || Failed syncs=${dailySyncFailCount}. Minimum successful sync percentage=${
+        MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE * 100
+      }%`
+    )
+  }
+
+  // Check historical sync data for rolling window 30 days
+  const { thirtyDayRollingSyncSuccessCount, thirtyDayRollingSyncFailCount } =
+    verboseHealthCheckResp
+  if (
+    thirtyDayRollingSyncSuccessCount &&
+    thirtyDayRollingSyncFailCount &&
+    thirtyDayRollingSyncSuccessCount + thirtyDayRollingSyncFailCount >
+      MINIMUM_ROLLING_SYNC_COUNT &&
+    thirtyDayRollingSyncSuccessCount /
+      (thirtyDayRollingSyncFailCount + thirtyDayRollingSyncSuccessCount) <
+      MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE
+  ) {
+    throw new Error(
+      `Rolling sync data shows that this node fails at a high rate of syncs. Successful syncs=${thirtyDayRollingSyncSuccessCount} || Failed syncs=${thirtyDayRollingSyncFailCount}. Minimum successful sync percentage=${
+        MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE * 100
+      }%`
+    )
+  }
+}
+
+/**
+ * @param node the endpoint to check for health
+ * @param isPrimary whether or not the node is a primary (primaries have a longer grace period)
+ * @return true if the node is either:
+ * - healthy; or
+ * - unhealthy but in a grace period (extra time where it can be unhealthy before being reconfiged)
+ */
+async function isNodeHealthyOrInGracePeriod(node: string, isPrimary: boolean) {
```

Comment:
this is renamed from isPrimaryHealthy in the JS version and changed a bit to support secondaries having a grace period as well